access_request_handler.go

```diff
@@ -80,14 +79,20 @@ func (f *Fosite) NewAccessRequest(ctx context.Context, r *http.Request, session
 		return accessRequest, errorsx.WithStack(ErrInvalidRequest.WithHint("Request parameter 'grant_type' is missing"))
 	}
 
-	client, err := f.AuthenticateClient(ctx, r, r.PostForm)
-	if err != nil {
-		return accessRequest, err
-	}
-	accessRequest.Client = client
-
 	var found = false
+	var isClientAuth = false
 	for _, loader := range f.TokenEndpointHandlers {
+		if !loader.CanHandleTokenEndpointRequest(accessRequest) {
+			continue
+		}
+		if !loader.CanSkipClientAuth(accessRequest) && !isClientAuth {
```

Comment:
I think I understand what this does, but it looks a bit dangerous to me because one could be able to skip client authentication in a future regression. Imagine a misplaced `continue` statement, for example.

Could we maybe solve this another way? One such idea would be to always authenticate the client (as before), but if an error occurs, we don't return it immediately but instead iterate through all `TokenEndpointHandlers` which have `CanSkipClientAuth()`.

```
 	client, clientErr := f.AuthenticateClient(ctx, r, r.PostForm)

 	var found = false
 	for _, loader := range f.TokenEndpointHandlers {
 		if !loader.CanHandleTokenEndpointRequest(accessRequest) {
 			continue
 		}
		if !loader.CanSkipClientAuth(accessRequest) && clientErr != nil {
 			continue
 		}
		if err := loader.HandleTokenEndpointRequest(ctx, accessRequest); err == nil {
 			found = true
		// ...
        }

	if !found && clientErr != nil {
		return nil, errorsx.WithStack(clientErr)
	}
```

An alternative would be to use `CanHandleTokenEndpointRequest` and check if `accessRequest.Client` is set - although I think that is also pretty unsafe.

Reply:
Changed it towards your suggestion, have a look, thanks!